pkg/sql/sqlbase/type_desc.go

```diff
@@ -188,6 +190,100 @@ func (desc *MutableTypeDescriptor) Immutable() DescriptorInterface {
 	return NewImmutableTypeDescriptor(*protoutil.Clone(desc.TypeDesc()).(*TypeDescriptor))
 }
 
+// EnumMembers is a sortable list of TypeDescriptor_EnumMember, sorted by the
+// physical representation.
+type EnumMembers []TypeDescriptor_EnumMember
+
+func (e EnumMembers) Len() int { return len(e) }
+func (e EnumMembers) Less(i, j int) bool {
+	return bytes.Compare(e[i].PhysicalRepresentation, e[j].PhysicalRepresentation) < 0
+}
+func (e EnumMembers) Swap(i, j int) { e[i], e[j] = e[j], e[i] }
+
+// Validate performs validation on the TypeDescriptor.
+func (desc *TypeDescriptor) Validate(ctx context.Context, txn *kv.Txn, codec keys.SQLCodec) error {
+	// Validate local properties of the descriptor.
+	if err := validateName(desc.Name, "type"); err != nil {
+		return err
+	}
+
+	if desc.ID == InvalidID {
+		return errors.AssertionFailedf("invalid ID %d", errors.Safe(desc.ID))
+	}
+	if desc.ParentID == InvalidID {
+		return errors.AssertionFailedf("invalid parentID %d", errors.Safe(desc.ParentID))
+	}
+
+	switch desc.Kind {
+	case TypeDescriptor_ENUM:
+		// All of the enum members should be in sorted order.
+		if !sort.IsSorted(EnumMembers(desc.EnumMembers)) {
+			return errors.AssertionFailedf("enum members are not sorted %v", desc.EnumMembers)
+		}
+		// Ensure there are no duplicate enum physical reps.
+		for i := 0; i < len(desc.EnumMembers)-1; i++ {
+			if bytes.Equal(desc.EnumMembers[i].PhysicalRepresentation, desc.EnumMembers[i+1].PhysicalRepresentation) {
+				return errors.AssertionFailedf("duplicate enum physical rep %v", desc.EnumMembers[i].PhysicalRepresentation)
+			}
+		}
+		// Ensure there are no duplicate enum labels.
+		members := make(map[string]struct{}, len(desc.EnumMembers))
+		for i := range desc.EnumMembers {
+			_, ok := members[desc.EnumMembers[i].LogicalRepresentation]
+			if ok {
+				return errors.AssertionFailedf("duplicate enum member %q", desc.EnumMembers[i].LogicalRepresentation)
+			}
+			members[desc.EnumMembers[i].LogicalRepresentation] = struct{}{}
+		}
+	case TypeDescriptor_ALIAS:
+		if desc.Alias == nil {
+			return errors.AssertionFailedf("ALIAS type desc has nil alias type")
+		}
+	default:
+		return errors.AssertionFailedf("invalid desc kind %s", desc.Kind.String())
+	}
+
+	// Validate all cross references on the descriptor.
+
+	// Check that the parent DB and schema exist.
+	{
+		res, err := txn.Get(ctx, MakeDescMetadataKey(codec, desc.ParentID))
```

Comment:
I know that the other validate method doesn't do this either, but I think we should switch the APIs of these to use a non concrete txn, so that we can pass in an object that caches some of these lookups. Maybe it's for another commit, but just wanted to put it out there.

Reply:
Yeah, I'll add that to the validate meta issue. Changing it to a `protogetter` is probably best.